### DISCUSSION TITLE
Warp math block into div block using pandoc filter

### DISCUSSION THREAD

**Discussion by Xmarmalade** [jgm/pandoc#9870] [Q&A] [answered]
I am using pandoc to convert markdown files to docx files, but I need to apply a separate format to the math block in markdown.
I found that there is no standalone format for math block in pandoc's default reference docx file, so I added a custom formatting like the following:

```
<div custom-style="Equations">

$$\begin{aligned}
   V_{\mathrm{yFk}}&=(1.08-1)\times7.186\times0.4m_{\mathrm{r}}\\
   &=0.230m_{\mathrm{r}}
\end{aligned}$$

</div>
```
the AST should like this
```
[ Para
      [ Math
          DisplayMath
          "\\begin{aligned}\n   V_{\\mathrm{yFk}}&=(1.08-1)\\times7.186\\times0.4m_{\\mathrm{r}}\\\\\n   &=0.230m_{\\mathrm{r}}\n\\end{aligned}"
      ]
  , Div
      ( "" , [] , [ ( "custom-style" , "Equations" ) ] )
      [ Para
          [ Math
              DisplayMath
              "\\begin{aligned}\n   V_{\\mathrm{yFk}}&=(1.08-1)\\times7.186\\times0.4m_{\\mathrm{r}}\\\\\n   &=0.230m_{\\mathrm{r}}\n\\end{aligned}"
          ]
      ]
  ]
```
However, this seems to be a pain as I have hundreds of math blocks to add, is there any way to write a filter or other ways to add custom style div?

**Comment by cagix** [comment]
you could try something like

```lua
function Math(el)
    if el.mathtype:match 'DisplayMath' then
        local kv = {}
        kv['custom-style'] = 'Equations'
        return pandoc.Div(el, kv)
    end
end
```

(not tested)

**Reply by cagix** [reply]
on second thought, `Math` seems to be _inline_, and `Div` wants a _block_, so maybe the last line should read `return pandoc.Div({el}, kv)`?

**Reply by cagix** [reply]
on third tought, since `Math` is _inline_, we can't return a _block_ ... so either you need to run over all blocks to find block math inside or you could try `RawInline` like

```lua
function Math(el)
    if el.mathtype:match 'DisplayMath' then
        return {
            pandoc.RawInline("html", "<div custom-style='Equations'>"),
            el,
            pandoc.RawInline("html", "</div>")
        }
    end
end
```

**Reply by Xmarmalade** [reply]
> on third tought, since `Math` is _inline_, we can't return a _block_ ... so either you need to run over all blocks to find block math inside or you could try `RawInline` like在第三个 tought 上，由于 `Math` 是内联的，我们不能返回一个块......因此，要么您需要遍历所有块才能在里面找到块数学，要么您可以尝试 `RawInline` 像
> 
> ```lua
> function Math(el)
>     if el.mathtype:match 'DisplayMath' then
>         return {
>             pandoc.RawInline("html", "<div custom-style='Equations'>"),
>             el,
>             pandoc.RawInline("html", "</div>")
>         }
>     end
> end
> ```

This seems to work, but MS Word doesn't recognize them correctly as 'Equations', the AST looks like

```
Para
      [ RawInline (Format "html") "<div custom-style='Equations'>"
      , Math
          DisplayMath
          "\\begin{aligned}\n   V_{\\mathrm{yFk}}&=(1.08-1)\\times7.186\\times0.4m_{\\mathrm{r}}\\\\\n   &=0.230m_{\\mathrm{r}}\n\\end{aligned}"
      , RawInline (Format "html") "</div>"
      ]
```
in MS word, the 'Equations' blocks still have 'Body Text' style

and im trying using python filter to check every para, but i dont know how to put the result together

```python
def warp_math_in_div(key, value, format, meta):
    if key == 'Para':
        if value[0]['t'] == 'Math':
          # dont know return <div custom-style="Equations">$$math$$</div>
          return Div(...)
 ```

**Reply by jgm** [reply]
Changing the last line of the first attempt above to something like this should give you what you want:
```lua
        return pandoc.Div(pandoc.Para({el}), kv)
```

**Reply by cagix** [reply]
> Changing the last line of the first attempt above to something like this should give you what you want:
> 
> ```lua
>         return pandoc.Div(pandoc.Para({el}), kv)
> ```

Yeah, that was my second thought. Unfortunately, `Math` is an inline element, so when returning a `Div` instead, Lua gives me an error.


```
% pandoc -L math.lua mwe.md  
Error running filter math.lua:
Inline, list of Inlines, or string expected, got Block
```

**Reply by cagix** [reply]
Seems, we have to take the scenic route:

```
result = empty list
for all blocks:
  if block contains block math:
    block_wo_math, math = remove_block_math(block)
    result.append(math)
    result.append(block_wo_math)
  else:
    result.append(block)
return result
```

**Reply by jgm** [reply]
Ah right, you'll have to split the paragraph and create a new block where the Math element was...

**Comment by Xmarmalade** [comment, marked as answer]
Found this solution:
```lua
function Para(para)
    -- if para sub el is math, then create a div with custom-style = Equations and put the para in it
    if para.content[1].t == 'Math' then
        local kv = {}
        kv['custom-style'] = 'Equations'
        return pandoc.Div(para, kv)
    end
end
```
Before
```
Pandoc
  Meta { unMeta = fromList [] }
  [ Para
      [ Math
          DisplayMath
          "\\begin{aligned}\n   V_{\\mathrm{yFk}}&=(1.08-1)\\times7.186\\times0.4m_{\\mathrm{r}}\\\\\n   &=0.230m_{\\mathrm{r}}\n\\end{aligned}"
      ]
  ]
```
After
```
Pandoc
  Meta { unMeta = fromList [] }
  [ Div
      ( "" , [] , [ ( "custom-style" , "Equations" ) ] )
      [ Para
          [ Math
              DisplayMath
              "\\begin{aligned}\n   V_{\\mathrm{yFk}}&=(1.08-1)\\times7.186\\times0.4m_{\\mathrm{r}}\\\\\n   &=0.230m_{\\mathrm{r}}\n\\end{aligned}"
          ]
      ]
  ]
```
Works well on my markdown, thanks!

**Reply by cagix** [reply]
that's a nice solution! however, keep in mind that this only works if the block math is (always) the first element in a paragraph, i.e. if there is always an empty line before each block math element. 

to make it work in the general case, you would have to implement as described above ([here](https://github.com/jgm/pandoc/discussions/9870#discussioncomment-9734987), [here](https://github.com/jgm/pandoc/discussions/9870#discussioncomment-9735113)).
